runners/google-cloud-dataflow-java/worker/src/test/java/org/apache/beam/runners/dataflow/worker/WindmillStateReaderTest.java

```diff
@@ -99,6 +113,713 @@ private ByteString intData(int value) throws IOException {
+    Mockito.verifyNoMoreInteractions(mockWindmill);
+
+    Windmill.KeyedGetDataRequest.Builder expectedRequest =
+        Windmill.KeyedGetDataRequest.newBuilder()
+            .setKey(DATA_KEY)
+            .setShardingKey(SHARDING_KEY)
+            .setWorkToken(WORK_TOKEN)
+            .setMaxBytes(WindmillStateReader.MAX_KEY_BYTES)
+            .addMultimapsToFetch(
+                Windmill.TagMultimapFetchRequest.newBuilder()
+                    .setTag(STATE_KEY_1)
+                    .setStateFamily(STATE_FAMILY)
+                    .setFetchEntryNamesOnly(false)
+                    .setFetchMaxBytes(WindmillStateReader.INITIAL_MAX_MULTIMAP_BYTES));
+
+    Windmill.KeyedGetDataResponse.Builder response =
+        Windmill.KeyedGetDataResponse.newBuilder()
+            .setKey(DATA_KEY)
+            .addTagMultimaps(
+                Windmill.TagMultimapFetchResponse.newBuilder()
+                    .setTag(STATE_KEY_1)
+                    .setStateFamily(STATE_FAMILY)
+                    .addEntries(
+                        Windmill.TagMultimapEntry.newBuilder()
+                            .setEntryName(STATE_MULTIMAP_KEY_1)
+                            .addValues(intData(1))
+                            .addValues(intData(2)))
+                    .addEntries(
+                        Windmill.TagMultimapEntry.newBuilder()
+                            .setEntryName(STATE_MULTIMAP_KEY_2)
+                            .addValues(intData(10))
+                            .addValues(intData(20))));
+    Mockito.when(mockWindmill.getStateData(COMPUTATION, expectedRequest.build()))
+        .thenReturn(response.build());
+
+    Iterable<Map.Entry<ByteString, Iterable<Integer>>> results = future.get();
+    Mockito.verify(mockWindmill).getStateData(COMPUTATION, expectedRequest.build());
+    int foundEntries = 0;
+    for (Map.Entry<ByteString, Iterable<Integer>> entry : results) {
+      if (entry.getKey().equals(STATE_MULTIMAP_KEY_1)) {
+        foundEntries++;
+        assertThat(entry.getValue(), Matchers.containsInAnyOrder(1, 2));
+      } else {
+        foundEntries++;
+        assertEquals(STATE_MULTIMAP_KEY_2, entry.getKey());
+        assertThat(entry.getValue(), Matchers.containsInAnyOrder(10, 20));
+      }
+    }
+    assertEquals(2, foundEntries);
+    Mockito.verifyNoMoreInteractions(mockWindmill);
+    assertNoReader(future);
+  }
+
+  private static void assertMultimapEntries(
+      Iterable<Map.Entry<ByteString, Iterable<Integer>>> expected,
+      List<Map.Entry<ByteString, List<Integer>>> actual) {
+    Map<ByteString, List<Integer>> expectedMap = Maps.newHashMap();
+    for (Map.Entry<ByteString, Iterable<Integer>> entry : expected) {
+      ByteString key = entry.getKey();
+      if (!expectedMap.containsKey(key)) expectedMap.put(key, new ArrayList<>());
+      entry.getValue().forEach(expectedMap.get(key)::add);
+    }
+    for (Map.Entry<ByteString, List<Integer>> entry : actual) {
+      assertThat(
```

Comment:
explicit check that expectedMap contains the key?
to avoid weird possiblities that returning null from remove possibly matches some weird actual state.